src/theory/fp/theory_fp_type_rules.cpp

```diff
@@ -66,51 +85,79 @@ TypeNode RoundingModeConstantTypeRule::computeType(NodeManager* nodeManager,
   return nodeManager->roundingModeType();
 }
 
+TypeNode FloatingPointFPTypeRule::preComputeType(NodeManager* nm, TNode n)
+{
+  return TypeNode::null();
+}
 TypeNode FloatingPointFPTypeRule::computeType(NodeManager* nodeManager,
                                               TNode n,
                                               bool check,
                                               std::ostream* errOut)
 {
   TRACE("FloatingPointFPTypeRule");
 
-  TypeNode signType = n[0].getType(check);
-  TypeNode exponentType = n[1].getType(check);
-  TypeNode significandType = n[2].getType(check);
+  TypeNode signType = n[0].getTypeOrNull();
+  TypeNode exponentType = n[1].getTypeOrNull();
+  TypeNode significandType = n[2].getTypeOrNull();
 
-  if (!signType.isBitVector() || !exponentType.isBitVector()
-      || !significandType.isBitVector())
+  if (!signType.isMaybeKind(kind::BITVECTOR_TYPE)
+      || !exponentType.isMaybeKind(kind::BITVECTOR_TYPE)
+      || !significandType.isMaybeKind(kind::BITVECTOR_TYPE))
   {
-    throw TypeCheckingExceptionPrivate(n,
-                                       "arguments to fp must be bit vectors");
+    if (errOut)
+    {
+      (*errOut) << "arguments to fp must be bit vectors";
+    }
+    return TypeNode::null();
+  }
+  // if not concrete, we are abstract floating point
+  if (!exponentType.isBitVector() || !significandType.isBitVector())
+  {
+    return nodeManager->mkAbstractType(kind::FLOATINGPOINT_TYPE);
   }
-
-  uint32_t signBits = signType.getBitVectorSize();
   uint32_t exponentBits = exponentType.getBitVectorSize();
   uint32_t significandBits = significandType.getBitVectorSize();
 
   if (check)
   {
-    if (signBits != 1)
+    if (signType.isBitVector())
```

Comment:
Shouldn't the sign always be a bit-vector?

Reply:
Not if we allow gradual types.  It could be `?`, `?BitVec` or `(_ BitVec n)` in this case.  This checks if it is the latter case.

Reply:
See my comment down below, either you have to check for the "invalid" condition in this `if`, or you have to remove the `else` below.

Reply:
Yes, good catch, I've removed the else.